src/ndt/test_c2s_impl.hpp

```diff
@@ -0,0 +1,193 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+#ifndef SRC_NDT_TEST_C2S_IMPL_HPP
+#define SRC_NDT_TEST_C2S_IMPL_HPP
+
+#include "src/ndt/internal.hpp"
+
+namespace mk {
+namespace ndt {
+namespace test_c2s {
+
+template <MK_MOCK_NAMESPACE(net, connect)>
+void coroutine_impl(std::string address, int port, double runtime,
+                    Callback<Error, Continuation<Error>> cb, double timeout,
+                    Settings settings, Var<Logger> logger,
+                    Var<Reactor> reactor) {
+
+    // Performance note: This implementation does some string copies
+    // when sending but in localhost testing this does not seem to be
+    // a problem when sending against a MK based discard server.
+    //
+    // Otherwise, a possible improvement would be to have a fixed
+    // size buffer and bypass bufferevent's output logic.
+    //
+    // In my tests (where speed of course depends on the computer you
+    // use), I was able to send at over 1.1 GiB/s and similar speeds
+    // were achieved using stripped down NDT code from `test_c2s_clt.c`.
+
+    dump_settings(settings, "ndt/c2s", logger);
+
+    std::string str = random_printable(8192);
+
+    logger->debug("ndt: connect ...");
+    net_connect(address, port,
+                [=](Error err, Var<Transport> txp) {
+                    logger->debug("ndt: connect ... %d", (int)err);
+                    if (err) {
+                        cb(err, nullptr);
+                        return;
+                    }
+                    logger->info("Connected to %s:%d", address.c_str(), port);
+                    logger->debug("ndt: suspend coroutine");
+                    cb(NoError(), [=](Callback<Error> cb) {
+                        double begin = time_now();
+                        Var<double> previous(new double(begin));
+                        Var<size_t> count(new size_t(0));
+                        logger->debug("ndt: resume coroutine");
+                        logger->info("Starting upload");
+                        txp->set_timeout(timeout);
+                        txp->on_flush([=]() {
+                            double now = time_now();
+                            if (now - *previous > 0.5) {
+                                double x =
+                                    (*count * 8) / 1000 / (now - *previous);
```

Comment:
Count is the result of calling size() on string, which is the size of the string in bytes so there is no need to multiply this by 8.